keyboards/handwired/aball/adns.c

```diff
@@ -0,0 +1,263 @@
+    // bit 1 = 1 for 3k mode, other bits are reserved
+    adns_write(REG_Configuration_IV, 0x02);
+
+    // write 0x1d in SROM_enable reg for initializing
+    adns_write(REG_SROM_Enable, 0x1d);
+
+    // wait for more than one frame period
+    // assume that the frame rate is as low as 100fps... even if it should never be that low
+    wait_ms(10);
+
+    // write 0x18 to SROM_enable to start SROM download
+    adns_write(REG_SROM_Enable, 0x18);
+
+    // write the SROM file (=firmware data)
+
+    // write burst destination adress
+    adns_begin();
+    spi_write(REG_SROM_Load_Burst | 0x80);
+    wait_us(15);
+    // send all bytes of the firmware
+    unsigned char c;
+    for(int i = 0; i < firmware_length; i++){
+        c = (unsigned char)pgm_read_byte(firmware_data + i);
+        spi_write(c);
+        wait_us(15);
+    }
+
+    adns_end();
+    wait_ms(10);
+
+    // enable laser(bit 0 = 0b), in normal mode (bits 3,2,1 = 000b)
+    // reading the actual value of the register is important because the real
+    // default value is different from what is said in the datasheet, and if you
+    // change the reserved bytes (like by writing 0x00...) it would not work.
+    uint8_t laser_ctrl0 = adns_read(REG_LASER_CTRL0);
+    adns_write(REG_LASER_CTRL0, laser_ctrl0 & 0xf0);
+
+    wait_ms(1);
+
+    // set the configuration_I register to set the CPI
+    // 0x01 = 50, minimum
+    // 0x44 = 3400, default
+    // 0x8e = 7100
+    // 0xA4 = 8200, maximum
+    adns_write(REG_Configuration_I, 0x08);
+
+    wait_ms(100);
+    dprint("INIT ENDED\n");
+}
+
+int16_t convertDeltaToInt(uint8_t high, uint8_t low){
+
+    // join bytes into twos compliment
+    uint16_t twos_comp = (high << 8) | low;
+
+    // convert twos comp to int
+    if (twos_comp & 0x8000)
+        return -1 * ((twos_comp ^ 0xffff) + 1);
+
+    return twos_comp;
+}
+
+void readSensor(void) {
+
+    adns_begin();
+
+    // send adress of the register, with MSBit = 1 to indicate it's a write
+    spi_write(REG_Motion_Burst & 0x7f);
+
+    uint8_t burst_data[pixel_sum];
+
+    for (int i = 0; i < pixel_sum; ++i) {
+        burst_data[i] = spi_read();
+    }
+
+    delta_x += convertDeltaToInt(burst_data[delta_x_h], burst_data[delta_x_l]);
+    delta_y += convertDeltaToInt(burst_data[delta_y_h], burst_data[delta_y_l]);
+    motion_ind = burst_data[motion];
+    adns_end();
+
+}
+void pointing_device_task(void) {
+    readSensor();
+    report_mouse_t report = pointing_device_get_report();
+
+    //if(delta_x + delta_y > 0)
+    //uprintf("M: %d\tX: %d\tY: %d\n", motion_ind, delta_x, delta_y);
+    // clamp deltas from -127 to 127
+    report.x = delta_x < -127 ? 127 : delta_x > 127 ? 127 : delta_x;
+    report.x = -report.x;
+
+    report.y = delta_y < -127 ? 127 : delta_y > 127 ? 127 : delta_y;
+
+    // reset deltas
+    delta_x = 0;
+    delta_y = 0;
+
+    pointing_device_set_report(report);
+    pointing_device_send();
```

Comment:
Just a heads up, without any check here about changes in the report, this will keep the system awake.

This is an issue I ran into with the pimironi trackball module.  
If you're polling, it will cause this, but if you're using interrupts, it shouldn't be an issue.

Reply:
Also, the other issue here is that there is no way to customize the behavior (such as holding down a switch to switch to a "scrolling" mode, for instance)

Reply:
Thank you @drashna!   The heads up about keeping the machine awake makes sense as it keeps getting mouse reports. I've wrapped this so that it only occurs on change.

As for the customization... This is a bit of a strange beast, it's a trackball with NO buttons.. just a ball.  I think you have a Kyria, so you may appreciate my inspiration here:
https://github.com/brickbots/aball

Reply:
Welcome! Personally I just check if any change has occurred in the report, and not send if no change has occurred. Either way works, I think.

And that's nice! (And yup, I have a kyria)